libbpf-tools/ttysnoop.bpf.c

```diff
@@ -0,0 +1,145 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/* Copyright (c) 2021 Hengqi Chen */
+#include <vmlinux.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include "ttysnoop.h"
+
+#define WRITE			1
+/**
+ * starting from linux v5.14-rc1, ITER_IOVEC has value 0 instead of 4
+ * see https://github.com/torvalds/linux/blob/v5.14-rc1/include/linux/uio.h#L22
+ */
+#define ITER_IOVEC_v514		0
+
+/**
+ * starting from linux v5.14-rc1, struct iov_iter has new definition
+ * define our own struct iov_iter since vmlinux.h is outdated
+ * see https://github.com/torvalds/linux/blob/v5.14-rc1/include/linux/uio.h#L30-L50
+ */
+struct iov_iter___v514 {
+	__u8 iter_type;
+	bool data_source;
+	union {
+		const struct kvec *kvec;
+	};
+};
+
+static int zero = 0;
+const volatile __u64 pts = 0;
+const volatile int read_bytes = 256;
+const volatile int read_count = 16;
+
+extern __u32 LINUX_KERNEL_VERSION __kconfig;
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__uint(max_entries, 1);
+	__type(key, int);
+	__type(value, struct event);
+} heap SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(__u32));
+	__uint(value_size, sizeof(__u32));
+} events SEC(".maps");
+
+static int probe_tty_write(void *ctx, const char *buf, size_t count)
+{
+	struct event *eventp;
+	int i;
+
+	eventp = bpf_map_lookup_elem(&heap, &zero);
+	if (!eventp)
+		return 0;
+
+	for (i = 0; i < read_count; i++) {
+		if (bpf_probe_read_user(eventp->buf, BUF_SIZE, buf))
+			return 0;
+
+		if (count < BUF_SIZE) {
+			eventp->count = count;
+			eventp->buf[eventp->count] = '\0';
+		} else {
+			eventp->count = BUF_SIZE - 1;
+			eventp->buf[eventp->count] = '\0';
```

Comment:
nit: this `= '\0'` happens in the other branch as well, can take it out of the if